### ISSUE TITLE
Indentation checker does not check for single space indentation

### ISSUE THREAD

**Issue by PatrickMcSweeny** [exercism/ruby-analyzer#107]
I created a test locally to see if the indentation checker would approve a solution which used single space indentation, and it did approve the solution. This is because the indentation checker only checks to see if the difference in spaces between a line and the previous line is greater than two. A better solution would be to see if the difference in spaces is equal to zero or two. We could probably get even fancier using some of the functionality in the Rubocop AST library.

I would like to work on a PR to address this as well as the refactoring that is suggested [here](https://github.com/exercism/ruby-analyzer/blob/main/lib/analyzers/solution_representation.rb#L14).

**Comment by kotp** [comment]
I said "is not really wanted" but I think we will come to the agreement that it is wanted.  I want to be clear that I am not discouraging the work suggested here. Only that we could discuss it to be clear what the goal is.

That said, I think if it is detected that 1 space indentation is here, that we would report that it is a well accepted convention in Ruby code to use two spaces for indentation.  Not tab or one space, but two spaces.  It it would be related to write an analyzer that detects trailing white space as well.  Even lack of EOL on what should be the last line.

So we have plenty of areas that could be pointed out in the analyzer.

**Comment by PatrickMcSweeny** [comment]
> That said, since we are here, what is your refactoring idea for the second part?

I mistakenly linked to the wrong line, but there is a comment in the indentation checker saying that the blank lines could be stripped and then each_cons could be used to check two lines at a time. This would make for a more readable solution, and it would probably be better for performance as well.

**Comment by kotp** [comment]
> I mistakenly linked to the wrong line

Did you edit the post to correct the link for ease of navigation, discussion?  I did not notice it.  I will try to figure out  and then search for what you describe though.  I believe you mean [solution_representation.rb:+9](https://github.com/exercism/ruby-analyzer/blob/2b75af8d3b5962b48f9abed3a1ab18fc3972f428/lib/analyzers/solution_representation.rb#L9).  It is also the only other comment in the file that was linked to.

> This would make for a more readable solution, and it would probably be better for performance as well.

What do you have?

**Comment by PatrickMcSweeny** [comment]
https://github.com/PatrickMcSweeny/ruby-analyzer/commit/122e387cf1b1ee5aebadf59af67eb62fb5d65909

**Comment by kotp** [comment]
> [PatrickMcSweeny@0ab6329](https://github.com/PatrickMcSweeny/ruby-analyzer/commit/0ab6329462adec4588d2db55210ff57976241694)

```diff
diff --git a/lib/analyzers/solution_representation.rb b/lib/analyzers/solution_representation.rb
index efef5c7..c306eff 100644
--- a/lib/analyzers/solution_representation.rb
+++ b/lib/analyzers/solution_representation.rb
@@ -6,28 +6,13 @@ class SolutionRepresentation
     @code_to_analyze = code_to_analyze
   end
 
-  # REFACTOR: This could be refactored to strip blank
-  # lines and then use each_cons(2).
   def indentation_is_sensible?
-    previous_line = nil
-    code_to_analyze.lines.each do |line|
-      # If the previous line or this line is
-      # just a whitespace line, don't consider it
-      # when checking for indentation
-      unless previous_line.nil? ||
-             previous_line =~ /^\s*\n*$/ ||
-             line =~ /^\s*\n*$/
+    code_to_analyze.lines.reject { |line| line =~ /^\s*\n*$/ }.each_cons(2).all? do |lines|
+      line_1_space = lines.first[/^ */].size
+      line_2_space = lines.last[/^ */].size
 
-        previous_line_lspace = previous_line[/^ */].size
-        line_lspace = line[/^ */].size
-
-        return false if (previous_line_lspace - line_lspace).abs > 2
-      end
-
-      previous_line = line
+      (line_2_space - line_1_space).abs <= 2
     end
-
-    true
   end
 
   def has_target_module?
```

I see.

Which results in having:

```ruby
  def indentation_is_sensible?
    code_to_analyze.lines.reject { |line| line =~ /^\s*\n*$/ }.each_cons(2).all? do |lines|
      line_1_space = lines.first[/^ */].size
      line_2_space = lines.last[/^ */].size

      (line_2_space - line_1_space).abs <= 2
    end
  end
```

Rather than 

```ruby
  def indentation_is_sensible?
    previous_line = nil
    code_to_analyze.lines.each do |line|
      # If the previous line or this line is
      # just a whitespace line, don't consider it
      # when checking for indentation
      unless previous_line.nil? ||
             previous_line =~ /^\s*\n*$/ ||
             line =~ /^\s*\n*$/

        previous_line_lspace = previous_line[/^ */].size
        line_lspace = line[/^ */].size

        return false if (previous_line_lspace - line_lspace).abs > 2
      end

      previous_line = line
    end

    true
  end
```

Consider if we are only checking line 1 and 2.  The original has references that are relative, which makes sense to me, rather than specific line numbers, which does not seem to reflect what is happening here.  (Naming is hard, I am not sure what to name them, but partly because it is pulling my mind to line 1 and line 2, as it reads, rather than how it is probably intended.

---

Do we also have a related test to avoid regression of behavior?

**Comment by PatrickMcSweeny** [comment]
There are a series of indentation tests here: https://github.com/exercism/ruby-analyzer/blob/main/test/exercises/two_fer_test.rb#L321

**Comment by kotp** [comment]
> There are a series of indentation tests here: [exericsm/ruby-analyzer/two_fer_test.rb#L321]( https://github.com/exercism/ruby-analyzer/blob/main/test/exercises/two_fer_test.rb#L321)

Yes, but I was asking for the Leap Year exercise.  This is something that you would likely put together.  You can use the information you found to model that, though.

**Comment by PatrickMcSweeny** [comment]
The leap year analyzer does not currently check for indentation because it does not auto-approve submissions.

**Comment by kotp** [comment]
I see what you are saying.  I was asking about tests, if they needed to be changed.  I just ran the tests against your changes, and they still pass, so from what I can see, nothing broke, but I am doubting that the title of this issue is addressed via testing, since currently no single indentation is checked, and this change is meant to check for single space indentation.  Should we add a test to help prevent regressions?

I like the code change.  I think it is easier to read, addresses the comment concern, but would ask that names are considered for making things clear when we come back to read the code later.  `current_line` and `prior_line` might be names that I would use rather than `line_1`, `line_2` as part of the name, or whatever helps to make that code's intention clear.

To reiterate, if the current code does not check for a single space indentation, we likely want to make a test that ensures we do not have a regression, right?  A change without a test means that it would be easy to not see that some maintenance change reverts it back to not checking for the single space indentation.

**Comment by PatrickMcSweeny** [comment]
The commit I posted was a refactoring to improve readability and performance. It was not meant to check for single-space indentation. If you are interested in checking to make sure that indentation is done with two spaces, I could add that to my proposed change.

**Comment by PatrickMcSweeny** [comment]
This commit shows how I would check for two-space indentation:

https://github.com/PatrickMcSweeny/ruby-analyzer/commit/bc0bbd3d8d627eb0ab86cbb0e03c95df3c72a44b

**Comment by kotp** [comment]
> This commit shows how I would check for two-space indentation:
> 
> [PatrickMcSweeny@bc0bbd3](https://github.com/PatrickMcSweeny/ruby-analyzer/commit/bc0bbd3d8d627eb0ab86cbb0e03c95df3c72a44b)

For ease of reading:

```diff
diff --git c/lib/analyzers/solution_representation.rb w/lib/analyzers/solution_representation.rb
index efef5c7..2e9adb5 100644
--- c/lib/analyzers/solution_representation.rb
+++ w/lib/analyzers/solution_representation.rb
@@ -6,28 +6,13 @@ class SolutionRepresentation
     @code_to_analyze = code_to_analyze
   end
 
-  # REFACTOR: This could be refactored to strip blank
-  # lines and then use each_cons(2).
   def indentation_is_sensible?
-    previous_line = nil
-    code_to_analyze.lines.each do |line|
-      # If the previous line or this line is
-      # just a whitespace line, don't consider it
-      # when checking for indentation
-      unless previous_line.nil? ||
-             previous_line =~ /^\s*\n*$/ ||
-             line =~ /^\s*\n*$/
+    code_to_analyze.lines.reject { |line| line =~ /^\s*\n*$/ }.each_cons(2).all? do |lines|
+      line_1_space = lines.first[/^ */].size
+      line_2_space = lines.last[/^ */].size
 
-        previous_line_lspace = previous_line[/^ */].size
-        line_lspace = line[/^ */].size
-
-        return false if (previous_line_lspace - line_lspace).abs > 2
-      end
-
-      previous_line = line
+      [0, 2].include? (line_2_space - line_1_space).abs
     end
-
-    true
   end
 
   def has_target_module?
diff --git c/test/exercises/two_fer_test.rb w/test/exercises/two_fer_test.rb
index 3b321f8..652c827 100644
--- c/test/exercises/two_fer_test.rb
+++ w/test/exercises/two_fer_test.rb
@@ -336,7 +336,13 @@ class TwoFerTest < Minitest::Test
         "One for %s, one for me." % name
      end
     end
-  '].each.with_index do |source, idx|
+  ', '
+    class TwoFer
+     def self.two_fer(name="you")
+     "One for %s, one for me." % name
+     end
+    end'
+  ].each.with_index do |source, idx|
     define_method "test_incorrect_indentation_#{idx}" do
       # skip
       results = TwoFer::Analyze.(source)
```

**Comment by kotp** [comment]
Thanks for presenting.  This would normally happen in the forums, rather than here, and I hope the people that need to see this will get here from what we have posted on the forums.

I am reopening, as I believe this is the work we would want to see!  Feel free to open a pull request for this, and link to that PR from the discussion that is relevant in the forums, please.

**Comment by kotp** [comment]
While this is closed, I am not sure that the testing concern is fully answered.  It was auto-closed due to the associated PR being brought in.

Please do consider the testing being enough to cover so that if the related behavior is removed we will find out about it via testing.

**Comment by PatrickMcSweeny** [comment]
Is there anything else you would like to see covered testing-wise?

**Comment by kotp** [comment]
I thought there was an aspect of indentation to 1 space.  I think it was covered in the leap exercise, though.  I would not worry it too much, it will come when it comes.  I was only concerned that your code had changed behavior of (like more complete or specific checking) not only a refactor.
